Add Support for Restoring OCS Operator CRs in ODF CLI

OdedViner opened the pull request:
This PR adds support for restoring deleted Custom Resources (CRs) for the OCS operator via the ODF CLI. It includes logic for handling storageclusters CRs by dynamically setting groupName and versionResource, ensuring compatibility and reliability during the restore process.

OdedViner commented on cmd/odf/restore/crds.go:
```diff
@@ -20,7 +21,16 @@ var deletedCmd = &cobra.Command{
-		pkgrestore.RestoreCrd(cmd.Context(), root.ClientSets, root.OperatorNamespace, root.StorageClusterNamespace, args)
+		var groupName string
+		var versionResource string
+		if args[0] == "storageclusters" {
```
The command already includes argument validation through Cobra with the line:
```
Args: cobra.RangeArgs(1, 2),
```

This ensures that if the number of arguments is outside the 1–2 range, Cobra will return an error and prevent further execution. As shown in the output, passing three arguments results in:

```
$ ./bin/odf restore deleted  a b c
Error: accepts between 1 and 2 arg(s), received 3
```

OdedViner commented:
> With this change, we really should enable the restore of any of the ODF CRs, whether from Rook, OCS operator, CSI operator, Noobaa, etc.
> 
> One approach would be to just require `args[0]` to include the fully qualified CRD type such as `storageclusters.ocs.openshift.io`, then we just extract the group name from that full name. We could return an error if they passed some group name that we don't recognize as belonging to ODF.

Hi @travisn ,

I’ve implemented the change as suggested. The restore command now requires that the first argument be a fully qualified CRD type (for example, storageclusters.ocs.openshift.io). The code parses this input to extract the resource name, group, and API version, and it validates the group against a set of supported groups (Rook, OCS, CSI, NooBaa, etc.). If the provided group isn’t recognized as part of ODF, the command returns an error with a helpful message listing the supported groups.

This update should enable us to restore any of the ODF CRs regardless of the operator to which they belong. Please take a look at the changes and let me know if you have any further suggestions or feedback.

Thanks,
Oded

subhamkrai commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,53 @@
+	"storage.k8s.io":         "v1",
+	"odf.openshift.io":       "v1alpha1",
+	"noobaa.io":              "v1alpha1",
+	"csiaddons.openshift.io": "v1alpha1",
```
it is v1alpha1 please confirm and upate

OdedViner replied:
done

OdedViner commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,54 @@
+	"ocs.openshift.io":       "v1",
+	"ceph.rook.io":           "v1",
+	"storage.k8s.io":         "v1",
+	"csi.ceph.io":            "v1",
```
done

OdedViner commented on cmd/odf/restore/crds.go:
```diff
@@ -20,8 +62,19 @@ var deletedCmd = &cobra.Command{
+		// Parse the fully qualified CRD (e.g. "cephclusters.ceph.rook.io").
+		resourceName, groupName, version, err := parseFullyQualifiedCRD(args[0])
+		if err != nil {
+			fmt.Printf("Error parsing CRD type: %v\n", err)
```
done

OdedViner commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,54 @@
+var groupVersions = map[string]string{
+	"ocs.openshift.io":       "v1",
+	"ceph.rook.io":           "v1",
+	"storage.k8s.io":         "v1",
```
removed "storage.k8s.io" CRD

OdedViner commented:
This PR depends on the kubectl-rook-ceph PR https://github.com/rook/kubectl-rook-ceph/pull/354 that was merged today.
@subhamkrai @travisn , is there an option to run a bot to update the dependencies?

subhamkrai commented:
> This PR depends on the kubectl-rook-ceph PR [rook/kubectl-rook-ceph#354](https://github.com/rook/kubectl-rook-ceph/pull/354) that was merged today. @subhamkrai @travisn , is there an option to run a bot to update the dependencies?

That will require kubectl-rook-ceph plugin release which I'm not sure we are planning now but you can use kubectl-rook-ceph with commit that has your changes

OdedViner commented:
> > This PR depends on the kubectl-rook-ceph PR [rook/kubectl-rook-ceph#354](https://github.com/rook/kubectl-rook-ceph/pull/354) that was merged today. @subhamkrai @travisn , is there an option to run a bot to update the dependencies?
> 
> That will require kubectl-rook-ceph plugin release which I'm not sure we are planning now but you can use kubectl-rook-ceph with commit that has your changes

done

subhamkrai reviewed (commented):
Changes looks @OdedViner have you verified the changes are working as expected.

OdedViner commented:
When attempting to restore it storagecluster CR, the process encountered an error while trying to remove finalizers from the storageclusters CR
```
1.Get storagecluster
$ oc get storagecluster -n openshift-storage 
NAME                 AGE   PHASE   EXTERNAL   CREATED AT             VERSION
ocs-storagecluster   21h   Ready              2025-02-23T18:25:32Z   4.18.0

2.Delete Storagecluster
$ oc delete storagecluster ocs-storagecluster -n openshift-storage 
storagecluster.ocs.openshift.io "ocs-storagecluster" deleted

3.Verify storagecluster in deleteing state
$ oc get storagecluster -n openshift-storage 
NAME                 AGE   PHASE      EXTERNAL   CREATED AT             VERSION
ocs-storagecluster   21h   Deleting              2025-02-23T18:25:32Z   4.18.0

4.Run Restore CLI 
$ oc get crd | grep storageclusters
storageclusters.ocs.openshift.io                                  2025-02-23T18:19:26Z

$ ./bin/odf restore deleted storageclusters.ocs.openshift.io
Info: Detecting which resources to restore for crd "storageclusters"
Info: Restoring CR ocs-storagecluster
Warning: The resource ocs-storagecluster was found deleted. Do you want to restore it? yes | no

yes
Info: Proceeding with restoring deleting CR
Info: Scaling down the operator
Info: Deleting validating webhook rook-ceph-webhook if present
Info: Removing ownerreferences from resources with matching uid ad51048e-cad3-47ce-88a7-c599cd14869a
Info: Removing owner references for secret onboarding-private-key
Info: Removed ownerReference for Secret: onboarding-private-key

Info: Removing owner references for secret onboarding-ticket-key
Info: Removed ownerReference for Secret: onboarding-ticket-key

Info: Removing owner references for configmaps rook-config-override
Info: Removed ownerReference for configmap: rook-config-override

Info: Removing owner references for service ocs-metrics-exporter
Info: Removed ownerReference for service: ocs-metrics-exporter

Info: Removing owner references for deployment ocs-metrics-exporter
Info: Removed ownerReference for deployment: ocs-metrics-exporter

Info: Removing owner references for deployment rook-ceph-tools
Info: Removed ownerReference for deployment: rook-ceph-tools

Info: Removing finalizers from storageclusters/ocs-storagecluster
Error: Failed to update resource "ocs-storagecluster" for crd. the server could not find the requested resource
```

subhamkrai reviewed (commented):
Please squah your commits and did you tested the latest changes with odf cluster?

subhamkrai commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,63 @@
+}
+
+// keys returns the keys of a string map. It is used to print out supported group names.
+func keys(m map[string]string) []string {
```
```suggestion
func groupNameKeys(m map[string]string) []string {
```
Let's have name related to their work

OdedViner replied:
done

subhamkrai commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,63 @@
+	resourceName = parts[0]
+	groupName = parts[1]
+
+	ver, ok := groupVersions[groupName]
```
```suggestion
	version, ok := groupVersions[groupName]
```

OdedViner replied:
done

subhamkrai commented on cmd/odf/restore/crds.go:
```diff
@@ -20,8 +71,54 @@ var deletedCmd = &cobra.Command{
+			newArgs[1] = args[1]
+		}
+		var customResources []pkgrestore.CustomResource
+		if contains(newArgs, "storageclusters") {
```
```suggestion
		if contains(newArgs, "storageclusters") {
```
I think we can use `string.Slices` here without https://go.dev/play/p/sj50jslDBUY to make code clean

OdedViner replied:
done

subhamkrai reviewed (changes requested):
Apart from these nits, changes looks good to me

subhamkrai commented:
The changes look good to me. Thanks

@travisn, ready to merge?
If I approve, the bot will merge, so I have not put lgtm tag

subhamkrai commented on go.sum:
```diff
@@ -743,8 +743,8 @@ github.com/rogpeppe/go-internal v1.6.1/go.mod h1:xXDCJY+GAPziupqXw64V24skbSoqbTE
 github.com/rogpeppe/go-internal v1.12.0/go.mod h1:E+RYuTGaKKdloAfM02xzb0FW3Paa99yedzYV+kq4uf4=
-github.com/rook/kubectl-rook-ceph v0.9.3 h1:+7THA8a+S2ArJrs9jpY1eJscAjPBKjlLXSmCVPU3eoY=
-github.com/rook/kubectl-rook-ceph v0.9.3/go.mod h1:dOQ+Yccc41DxZqe9jpvAUHsYTquYP/SKClrPmG70SLM=
+github.com/rook/kubectl-rook-ceph v0.9.4-0.20250428051344-dbfe77cc57a1 h1:znvPe0apxkTdkdiVINk0DfUbMGt6Vv+I9mgKfsr3odY=
```
https://github.com/rook/kubectl-rook-ceph/releases/tag/v0.9.4

OdedViner replied:
done

OdedViner commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,55 @@
+var groupVersions = map[string]string{
+	"ocs.openshift.io":       "v1",
+	"ceph.rook.io":           "v1",
+	"csi.ceph.io":            "v1beta1",
```
done

leelavg commented on cmd/odf/restore/crds.go:
```diff
@@ -1,12 +1,55 @@
 )
 
+// groupVersions defines the supported CRD groups and their corresponding API versions.
+var groupVersions = map[string]string{
+	"ocs.openshift.io":       "v1",
+	"ceph.rook.io":           "v1",
+	"csi.ceph.io":            "v1alpha1",
+	"odf.openshift.io":       "v1alpha1",
+	"noobaa.io":              "v1alpha1",
+	"csiaddons.openshift.io": "v1alpha1",
+}
+
+// groupNameKeys returns the keys of a string map. It is used to print out supported group names.
+func groupNameKeys(m map[string]string) []string {
+	out := make([]string, 0, len(m))
+	for k := range m {
+		out = append(out, k)
+	}
+	return out
+}
+
+// parseFullyQualifiedCRD takes a fully qualified CRD type of the form "resource.group"
+// (for example, "cephclusters.ceph.rook.io") and returns the resource name, group name, and
+// the API version associated with that group. It returns an error if the format is invalid
+// or the group is not recognized.
+func parseFullyQualifiedCRD(fqcrd string) (resourceName, groupName, version string, err error) {
+	parts := strings.SplitN(fqcrd, ".", 2)
+	if len(parts) != 2 {
+		return "", "", "", fmt.Errorf("invalid CRD format %q; expected format <resource>.<group>", fqcrd)
+	}
+	resourceName = parts[0]
+	groupName = parts[1]
+
+	version, ok := groupVersions[groupName]
+	if !ok {
+		return "", "", "", fmt.Errorf("unsupported group %q; supported groups are: %v", groupName, groupNameKeys(groupVersions))
+	}
+	return resourceName, groupName, version, nil
+}
+
```
any reason in not using the established GVK/GVR schema from https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/schema/group_version.go or https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/group_version.go if we want a higher level of abstraction.

is it like you don't want to import Go Structs from other repos? as of now iiuc, we already made an effort moving APIs into a submodule.

OdedViner replied:
@leelavg The input format is `./bin/odf restore deleted storageclusters.ocs.openshift.io`, where only the CR name and group are provided (storageclusters.ocs.openshift.io). We do not specify the version. Please add the relevant code to improve the current implementation.

leelavg commented on cmd/odf/restore/crds.go:
```diff
@@ -20,8 +63,54 @@ var deletedCmd = &cobra.Command{
+		if slices.Contains(newArgs, "storageclusters") {
+			customResources = []pkgrestore.CustomResource{
+				// ceph.rook.io/v1
+				{Group: "ceph.rook.io", Version: "v1", Resource: "cephblockpoolradosnamespaces"},
```
how can we create a link back to the ocs-op to keep this list updated when a new resource is owned by storagecluster?

did you check that k8s primitives (secrets, configmaps etc) that are owned by storagecluster are ok to be deleted?

apart from ensuring pods are running what other checks are performed after restoration of storagecluster CR?

iirc, storagecluster UID is being used in DR related activities as well, maybe that also gets effected or we need to have a statement of not performing this on DR enabled setups?

OdedViner replied:
@leelavg I have not tested it on a DR setup or in Provider mode.

OdedViner commented:
> just dropping it here as I didn't look into the plugins that we were developing upto now (pardon), usually https://github.com/kubernetes/cli-runtime is preferred as a base when developing kubectl plugins as it abstracts api-resources, discovery if required, generic printers etc.

@leelavg This is a downstream project, so I believe it would be more relevant for the upstream project: https://github.com/rook/kubectl-rook-ceph.

subhamkrai commented:
We also need to scale down `odf-operator` when restoring storageCluster.

OdedViner commented:
> We also need to scale down `odf-operator` when restoring storageCluster.

@subhamkrai 
The `StorageCluster` is an OCS CR, and I don't think the odf-operator is relevant here. As I understand it, the `odf-operator` is responsible for installing all operators, including OCS, Rook, NooBaa, Ceph CS and etc..

There is also a `StorageSystem` CR managed by the `odf-operator`, which allows for deploying multiple StorageClusters within the same cluster. However, I don't think we're using it, and I don't believe it's relevant in this case because the StorageCluster is a child resource, not a parent. Therefore, we shouldn't need to delete the ownerReference.

Please correct me if I'm wrong.

subhamkrai commented:
> > We also need to scale down `odf-operator` when restoring storageCluster.
> 
> @subhamkrai The `StorageCluster` is an OCS CR, and I don't think the odf-operator is relevant here. As I understand it, the `odf-operator` is responsible for installing all operators, including OCS, Rook, NooBaa, Ceph CSI and etc..
> 
> There is also a `StorageSystem` CR managed by the `odf-operator`, which allows for deploying multiple StorageClusters within the same cluster. However, I don't think we're using it, and I don't believe it's relevant in this case because the StorageCluster is a child resource, not a parent. Therefore, we shouldn't need to delete the ownerReference.
> 
> Please correct me if I'm wrong.

Okay, I see scaling odf-operator is not required here.

agarwal-mudit commented:
/cherry-pick release-4.19
